Source/bmalloc/libpas/src/libpas/pas_probabilistic_guard_malloc_allocator.c

```diff
@@ -301,6 +309,31 @@ void pas_probabilistic_guard_malloc_initialize_pgm_as_enabled(void)
     pas_probabilistic_guard_malloc_counter = 0;
 }
 
+void pas_probabilistic_guard_malloc_remove_metadata(pas_ptr_hash_map_entry * entry)
+{
+    if (pgm_index <= MAX_PGM_HASH_ENTRIES)
+        pgm_vector[pgm_index++] = entry;
+    else
+        PAS_ASSERT(pgm_index <= MAX_PGM_HASH_ENTRIES);
+
+    /*
+     * Check if max PGM entries have been reached if so free respective metadata for all freed memory.
+     * Note the length of metadata liveness will depend on how long it takes to fill all 10 entries.
+     */
+    if (pgm_index == MAX_PGM_HASH_ENTRIES) {
```

Comment:
Instead of clearing all of them at once can we instead just always remove the last one when we are fill up the buffer.

Reply:
I was not sure if we remove only first or last entry so biasing certain entries by keeping metadata for those longer ? I am not very clear what should be the right approach. Can revise this after Perf results.

Reply:
+1 to removing one at a time, otherwise doesn't that mean there's a predictable cycle where PGM is basically disabled (since we cleared out all PGM metadata at once)? It seems better to me if we use a ring buffer or something here and just free up old entries as we allocate new ones.

Reply:
Revised design to circular buffer for bookkeeping the PGM metadata entries and removing one at a time to make space for new entries.